app/models/link_history.rb

```diff
@@ -0,0 +1,41 @@
+class LinkHistory < ApplicationRecord
+  belongs_to :link
+
+  def add_errors(errors)
+    remove_resolved_errors(errors)
+    errors.each { |error| add_error(error) }
+  end
+
+  def add_error(message)
+    return if error_exists?(message)
+
+    link_errors << {
+      message: message,
+      started_at: DateTime.now
+    }
+
+    save
+  end
+
+  def clear_errors
+    update!(link_errors: [])
+  end
+
+private
+
+  #remove_resolved_errors removes any resolved errors from link_errors
```

Comment:
Ah oops, I didn't mean to imply that we needed comments here, I meant we should put a comment about why warnings haven't been included in this class yet and rename `add_errors` to `update_errors` to make it clear it does more than just adding.